lib/puppet/provider/cs_location/pcs.rb

```diff
@@ -88,14 +124,83 @@ def score=(should)
     @property_hash[:score] = should
   end
 
+  def boolean=(should)
+    @property_hash[:boolean] = should
+  end
+
+  def rule=(should)
+    @property_hash[:rule] = should
+  end
+
   # Flush is triggered on anything that has been detected as being
   # modified in the location_hash.
   # It calls several pcs commands to make the resource look like the
   # params.
   def flush
     unless @property_hash.empty?
-      cmd = [ command(:pcs), 'constraint', 'location', 'add', @property_hash[:name], @property_hash[:primitive], @property_hash[:node_name], @property_hash[:score]]
+
+      doc = REXML::Document.new
+
+      # add a new rsc_location XML element for a node-based constraint 
+      if @property_hash[:node_name]
+        rsc_location = doc.add_element 'rsc_location', {
+          'id'    => "#{@property_hash[:name]}",
+          'node'  => "#{@property_hash[:node_name]}",
+          'rsc'   => "#{@property_hash[:primitive]}",
+          'score' => "#{@property_hash[:score]}",
+        }
+        cmd_action = [ '--modify', '--allow-create', ]
+      end
+
+      # add a new rsc_location XML element for a rule-based constraint 
+      unless @property_hash[:rule].empty?
+        rsc_location = doc.add_element 'rsc_location', {
+          'id'  => "#{@property_hash[:name]}",
+          'rsc' => "#{@property_hash[:primitive]}",
+        }
+
+        # if there are more than 1 expressions defined in the array
+        # a boolean attribute is required in the rule element.
+        # 'and' is specified if no boolean parameter is present in the 
+        # property_hash
+        if @property_hash[:rule].length > 1
+          if @property_hash[:boolean].nil?
+            boolean = 'and'
+          else
+            boolean = @property_hash[:boolean]
+          end
+          rsc_location.add_element 'rule', {
+            'boolean-op' => boolean,
+            'id'         => "#{@property_hash[:name]}-rule",
+            'score'      => "#{@property_hash[:score]}",
+          }
+        else
+          rsc_location.add_element 'rule', {
+            'id'    => "#{@property_hash[:name]}-rule",
+            'score' => "#{@property_hash[:score]}",
+          }
+        end
+
+        # Add each expression hash to the  expression XML element in order.
+        @property_hash[:rule].each_with_index { |expression, index|
+          expression['id'] = "#{@property_hash[:name]}-rule-expr-#{index+1}"
+          rsc_location.elements['rule'].add_element 'expression', expression
+        }
+
+        # if you remove an expression from the cib, the replace switch is required,
+        # as the expression you want to get rid of will be ignored when using the modify switch
+        if (@property_hash[:existing_rule] and @property_hash[:rule].length < @property_hash[:existing_rule].length)
```

Comment:
what is `:existing_rule`, and why can it be `nil`?
why are we storing both, `:rule`, and `:existing_rule` in the same `@property_hash` in the provider, rather than having puppet determine whether the two are `insync` through the type?
or am i really misunderstanding what's going on here?